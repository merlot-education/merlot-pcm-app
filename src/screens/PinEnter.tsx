import React, { useState, useEffect } from 'react'
import { Alert, Keyboard, SafeAreaView, StyleSheet, View } from 'react-native'
import { useTranslation } from 'react-i18next'
import ReactNativeBiometrics from 'react-native-biometrics'
import { TextInput } from '../components'
import Button, { ButtonType } from '../components/button/Button'
import { Colors } from '../theme/theme'
import { getValueKeychain } from '../utils/keychain'

interface PinEnterProps {
  setAuthenticated: React.Dispatch<React.SetStateAction<boolean>>
}

const style = StyleSheet.create({
  container: {
    backgroundColor: Colors.background,
    margin: 20,
  },
  btnContainer: {
    marginTop: 20,
  },
})

const PinEnter: React.FC<PinEnterProps> = ({ setAuthenticated }) => {
  const [pin, setPin] = useState('')
  const [biometricFailed, setBiometricFailed] = useState(false)
  const [biometricSensorAvailable, setBiometricSensorAvailable] =
    useState(false)

  useEffect(() => {
    ReactNativeBiometrics.isSensorAvailable().then(resultObject => {
      const { available, biometryType } = resultObject
      if (available && biometryType === ReactNativeBiometrics.Biometrics) {
        setBiometricSensorAvailable(true)
      } else {
        setBiometricFailed(true)
      }
    })
  })
  const { t } = useTranslation()
  const checkPin = async (pin: string) => {
    const keychainEntry = await getValueKeychain({
      service: 'passcode',
    })
    if (keychainEntry && pin === keychainEntry.password) {
      setAuthenticated(true)
    } else {
      Alert.alert(t('PinEnter.IncorrectPin'))
    }
  }

  const biometricEnable = () => {
    ReactNativeBiometrics.isSensorAvailable().then(resultObject => {
      const { available, biometryType } = resultObject
      if (available && biometryType === ReactNativeBiometrics.Biometrics) {
        ReactNativeBiometrics.simplePrompt({
          promptMessage: t('Biometric.BiometricConfirm'),
        })
          .then(resultObject => {
            const { success } = resultObject
            if (success) {
              setAuthenticated(true)
            } else {
              Alert.alert(t('Biometric.BiometricCancle'))
            }
          })
          .catch(() => {
            Alert.alert(t('Biometric.BiometricFailed'))
            setBiometricFailed(true)
          })
      } else {
        Alert.alert(t('Biometric.BiometricNotSupport'))
        setBiometricFailed(true)
      }
    })
  }

  return (
    <SafeAreaView style={[style.container]}>
      {biometricFailed && (
        <>
          <TextInput
            label={t('Global.EnterPin')}
            accessible
            accessibilityLabel={t('Global.EnterPin')}
            placeholder={t('Global.6DigitPin')}
            placeholderTextColor={Colors.lightGrey}
            autoFocus
            maxLength={6}
            keyboardType="numeric"
            secureTextEntry
            value={pin}
            onChangeText={(pin: string) => {
              setPin(pin.replace(/[^0-9]/g, ''))
              if (pin.length === 6) {
                Keyboard.dismiss()
              }
            }}
          />
          <Button
            title={t('Global.Submit')}
            buttonType={ButtonType.Primary}
            onPress={() => {
              Keyboard.dismiss()
              checkPin(pin)
            }}
          />
        </>
      )}
      <View style={style.btnContainer}>
        {biometricSensorAvailable && !biometricFailed && (
          <Button
            title={t('Biometric.Biometric')}
            buttonType={ButtonType.Primary}
            onPress={biometricEnable}
          />
        )}
      </View>
    </SafeAreaView>
  )
}

export default PinEnter
