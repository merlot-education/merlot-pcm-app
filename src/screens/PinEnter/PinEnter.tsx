import React, { useState, useEffect, useCallback } from 'react'
import { Alert, Keyboard, SafeAreaView, StyleSheet, Text } from 'react-native'
import { useTranslation } from 'react-i18next'
import ReactNativeBiometrics from 'react-native-biometrics'
import { StackScreenProps } from '@react-navigation/stack'
import { TextInput, Loader } from '../../components'
import Button, { ButtonType } from '../../components/button/Button'
import { ColorPallet, TextTheme } from '../../theme/theme'
import { OnboardingStackParams, Screens } from '../../types/navigators'
import { KeychainStorageKeys } from '../../constants'
import {
  checkIfSensorAvailable,
  removeOnboardingCompleteStage,
  createMD5HashFromString,
  getValueFromKeychain,
  showBiometricPrompt,
} from './PinEnter.utils'
import { warningToast } from '../../utils/toast'

type PinEnterProps = StackScreenProps<OnboardingStackParams, Screens.EnterPin>

const style = StyleSheet.create({
  container: {
    backgroundColor: ColorPallet.grayscale.white,
    margin: 20,
  },
  bodyText: {
    ...TextTheme.normal,
    flexShrink: 1,
  },
  verticalSpacer: {
    marginVertical: 20,
    textAlign: 'center',
  },
})

const PinEnter: React.FC<PinEnterProps> = ({ navigation, route }) => {
  const { initAgent, setAuthenticated } = route.params
  const [pin, setPin] = useState('')
  const [loginAttemtsFailed, setLoginAttemtsFailed] = useState(0)
  const [loading, setLoading] = useState(false)
  const { t } = useTranslation()

  const startAgent = useCallback(async () => {
    const [email, passphrase, passcode] = await Promise.all([
      new Promise(resolve => {
        resolve(getValueFromKeychain(KeychainStorageKeys.Email))
      }),
      new Promise(resolve => {
        resolve(getValueFromKeychain(KeychainStorageKeys.Passphrase))
      }),
      new Promise(resolve => {
        resolve(getValueFromKeychain(KeychainStorageKeys.Passcode))
      }),
    ])
    if (email && passphrase) {
      const rawValue = email + passphrase.password.replace(/ /g, '')
      const seedHash = createMD5HashFromString(rawValue)
      console.log('email, passcode, seedHash', email, passcode, seedHash)
      initAgent(email.password, passcode.password, seedHash)
    }
  }, [initAgent])

  const checkBiometricIfPresent = useCallback(async () => {
    const { available, biometryType } = await checkIfSensorAvailable()
    if (available && biometryType === ReactNativeBiometrics.Biometrics) {
      const { success, error } = await showBiometricPrompt()
      if (success) {
        setLoading(true)
        await startAgent()
        setLoading(false)
        setAuthenticated(true)
      } else if (error) {
        warningToast(error)
      } else {
        warningToast(t('Biometric.BiometricCancle'))
      }
    } else {
      warningToast(t('Biometric.BiometricNotSupport'))
    }
  }, [setAuthenticated, startAgent, t])

  useEffect(() => {
    checkBiometricIfPresent()
  }, [checkBiometricIfPresent])

  const checkPin = async (pin: string) => {
    const [passcode] = await Promise.all([
      new Promise(resolve => {
        resolve(getValueFromKeychain(KeychainStorageKeys.Passcode))
      }),
    ])
    if (pin === passcode.password) {
      setLoading(true)
      await startAgent()
      setAuthenticated(true)
      setLoading(false)
    } else {
      warningToast(t('PinEnter.IncorrectPin'))
      setLoginAttemtsFailed(loginAttemtsFailed + 1)
      if (loginAttemtsFailed === 5) {
        Alert.alert(t('Registration.RegisterAgain'))
        navigation.navigate(Screens.Registration, { forgotPin: false })
        await removeOnboardingCompleteStage()
      }
    }
  }

  return (
    <SafeAreaView style={[style.container]}>
      <Loader loading={loading} />
      <TextInput
        label={t('Global.EnterPin')}
        accessible
        accessibilityLabel={t('Global.EnterPin')}
        placeholder={t('Global.6DigitPin')}
        placeholderTextColor={ColorPallet.baseColors.lightGrey}
        maxLength={6}
        keyboardType="numeric"
        secureTextEntry
        value={pin}
        onChangeText={(pin: string) => {
          setPin(pin.replace(/[^0-9]/g, ''))
          if (pin.length === 6) {
            Keyboard.dismiss()
          }
        }}
      />
      <Text
        style={[style.bodyText, style.verticalSpacer]}
        onPress={() =>
          navigation.navigate(Screens.Registration, { forgotPin: true })
        }
      >
        {t('Global.ForgotPin')}
      </Text>
      <Button
        title={t('Global.Submit')}
        buttonType={ButtonType.Primary}
        onPress={() => {
          Keyboard.dismiss()
          checkPin(pin)
        }}
      />
    </SafeAreaView>
  )
}

export default PinEnter
