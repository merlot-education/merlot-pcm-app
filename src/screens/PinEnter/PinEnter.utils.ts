import AsyncStorage from '@react-native-async-storage/async-storage'
import ReactNativeBiometrics from 'react-native-biometrics'
import md5 from 'md5'
import i18next from 'i18next'
import { LocalStorageKeys } from '../../constants'
import { getValueKeychain } from '../../utils/keychain'
import api from '../../api'

export const getValueFromKeychain = async (key: string) => {
  const data = await getValueKeychain({
    service: key,
  })
  return data
}

export const registerUser = async (email: string, otpId: string) => {
  const data = await api.auth.register({
    email,
    otpId,
  })
  return data
}

export const checkIfSensorAvailable = async () => {
  const result = await ReactNativeBiometrics.isSensorAvailable()
  return result
}

export const showBiometricPrompt = async () => {
  const result = await ReactNativeBiometrics.simplePrompt({
    promptMessage: i18next.t('Biometric.BiometricConfirm'),
  })
  return result
}

export const createMD5HashFromString = (value: string) => {
  const hash = String(md5(value))
  return hash
}

export const removeOnboardingCompleteStage = async () => {
  await AsyncStorage.removeItem(LocalStorageKeys.OnboardingCompleteStage)
}
