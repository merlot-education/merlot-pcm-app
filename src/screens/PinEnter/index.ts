import PinEnter from './PinEnter';

export default PinEnter;
