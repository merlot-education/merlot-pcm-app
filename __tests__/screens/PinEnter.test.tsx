import React from 'react'
import { create } from 'react-test-renderer'
import PinEnter from '../../src/screens/PinEnter'

jest.mock('react-native-keychain', () => 'PinEnter')
jest.mock('react-native-config', () => 'PinEnter')

describe('PinEnter Screen', () => {
  it('PinEnter correctly', () => {
    const tree = create(<PinEnter />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
