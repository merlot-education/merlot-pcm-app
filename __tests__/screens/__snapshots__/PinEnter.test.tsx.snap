// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinEnter Screen PinEnter correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "#FFF",
        "margin": 20,
      },
    ]
  }
>
  <View
    style={
      Object {
        "marginVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#03028A",
          "fontSize": 14,
          "fontWeight": "bold",
          "marginBottom": 3,
        }
      }
    >
      Global.EnterPin
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderBottomColor": "#dddddd",
            "borderBottomWidth": 0.5,
            "flexDirection": "row",
            "height": 44.5,
            "marginBottom": 0,
            "marginLeft": 15,
            "marginTop": 0,
            "paddingRight": 15,
          },
          Object {
            "borderBottomWidth": 0.5,
          },
        ]
      }
    >
      <TextInput
        accessibilityLabel="Global.EnterPin"
        accessible={true}
        autoFocus={true}
        clearButtonMode="never"
        editable={true}
        keyboardType="numeric"
        labelPosition="left"
        maxLength={6}
        onBlur={[Function]}
        onChange={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Global.6DigitPin"
        placeholderTextColor="#D3D3D3"
        secureTextEntry={false}
        selectionColor="#03028A"
        style={
          Array [
            Object {
              "height": 35,
            },
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "flex": 1,
              "fontSize": 17,
            },
            null,
            Object {},
            Array [
              Object {
                "backgroundColor": "#FFF",
                "borderRadius": 4,
                "borderWidth": 2,
                "color": "#03028A",
                "fontSize": 16,
                "padding": 10,
              },
              false,
            ],
          ]
        }
        textAlign="left"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderRadius": 5,
          "borderWidth": 1,
          "justifyContent": "center",
        },
        Object {
          "height": 47,
          "paddingLeft": 15,
          "paddingRight": 15,
        },
        Object {
          "backgroundColor": "#108ee9",
          "borderColor": "#108ee9",
        },
        false,
        false,
        undefined,
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
            },
            Object {
              "color": "#ffffff",
            },
            false,
            false,
          ]
        }
      >
        Global.Submit
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 20,
      }
    }
  />
</RCTSafeAreaView>
`;
